ctapipe/calib/camera/gainselection.py

```diff
@@ -1,173 +1,110 @@
     """
-    the PMTs on some cameras have 2 gain channels. select one
-    according to a threshold.
-
-    Parameters:
-    -----------
-    waveforms: np.ndarray
-        Array of shape (N_gain, N_pix, N_samp)
-    threshold: float
-        threshold (in PE/sample) of when to switch to low-gain channel
-    select_by_sample: bool
-        if true, select only low-gain *samples* when high-gain is over
-        threshold
-
-    Returns
-    -------
-    tuple:
-        gain-selected intensity, boolean array of which channel was chosen
+    Possible gain channels
     """
-
-    # if we have 2 channels:
-    if waveforms.shape[0] == 2:
-        waveforms = np.squeeze(waveforms)
-        new_waveforms = waveforms[0].copy()
-
-        if select_by_sample:
-            # replace any samples that are above threshold with low-gain ones:
-            gain_mask = waveforms[0] > threshold
-            new_waveforms[gain_mask] = waveforms[1][gain_mask]
-        else:
-            # use entire low-gain waveform if any sample of high-gain
-            # waveform is above threshold
-            gain_mask = (waveforms[0] > threshold).any(axis=1)
-            new_waveforms[gain_mask] = waveforms[1][gain_mask]
-
-    elif waveforms.shape[0] == 1:
-        new_waveforms = np.squeeze(waveforms)
-        gain_mask = np.zeros_like(new_waveforms).astype(bool)
-
-    else:
-        raise ValueError("input waveforms has shape %s. not sure what to do "
-                         "with that.", waveforms.shape)
-
-    return new_waveforms, gain_mask
+    HIGH = 0
+    LOW = 1
 
 
-class GainSelector(Component, metaclass=ABCMeta):
+class GainSelector(Component):
     """
     Base class for algorithms that reduce a 2-gain-channel waveform to a
     single waveform.
     """
-    @abstractclassmethod
-    def select_gains(self, cam_id, multi_gain_waveform):
+
+    def __call__(self, waveforms):
         """
-        Takes an input waveform and cam_id  and performs gain selection
+        Reduce the waveform to a single gain channel
+
+        Parameters
+        ----------
+        waveforms : ndarray
+            Waveforms stored in a numpy array of shape
+            (n_chan, n_pix, n_samples).
 
         Returns
         -------
-        tuple(ndarray, ndarray):
-            (waveform, gain_mask), where the gain_mask is a boolean array of
-            which gain channel was used.
+        reduced_waveforms : ndarray
+            Waveform with a single channel
+            Shape: (n_pix, n_samples)
         """
-        pass
+        if waveforms.ndim == 2:  # Return if already gain selected
+            return waveforms
+        elif waveforms.ndim == 3:
+            n_channels, n_pixels, _ = waveforms.shape
+            if n_channels == 1:  # Reduce if already single channel
+                return waveforms[0]
+            else:
+                pixel_channel = self.select_channel(waveforms)
+                return waveforms[pixel_channel, np.arange(n_pixels)]
```

Comment:
Just a question, why not just 'return waveforms[pixel_channel, :]?

Reply:
Unfortunately is isn't as simple to index the array like that. `waveforms[pixel_channel, :]` returns an array of shape (n_pix, n_pix, n_samples).

Reply:
(When pixel_channel is an array of n_pix itself)

Reply:
Ah ok